actions/generate-schedule/scheduler/scheduler.go

```
@@ -0,0 +1,98 @@
+			StartDate: start.Format(DateFormat),
+			User:      s.userSource.NextUser(),
+		}},
+	}, stop)
+}
+
+// ExtendSchedule takes a previously generated schedule and extends it. The user rotation continues as normal from the
+// last shift in the schedule.
+func (s *Scheduler) ExtendSchedule(previousSchedule *schedule.Schedule, stop time.Time) (*schedule.Schedule, error) {
+	if err := validatePreviousShifts(previousSchedule.Shifts); err != nil {
+		return nil, fmt.Errorf("error validating previous shifts: %v", err)
+	}
+
+	newSchedule := &schedule.Schedule{
+		Shifts: previousSchedule.Shifts,
+	}
+
+	mostRecentShift := previousSchedule.Shifts[len(previousSchedule.Shifts)-1]
+	if err := s.userSource.StartAfter(mostRecentShift.User); err != nil {
+		return nil, fmt.Errorf("error finding previous shift owner (%v) in user source: %v", mostRecentShift.User, err)
+	}
+
+	for start := s.startTime(mostRecentShift); start.Before(stop); start = start.AddDate(0, 0, s.shiftDurationDays) {
+		newSchedule.Shifts = append(newSchedule.Shifts, &schedule.Shift{
+			User:      s.userSource.NextUser(),
+			StartDate: start.Format(DateFormat),
+		})
+	}
+
+	return newSchedule, nil
+}
+
+func validatePreviousShifts(previousShifts []*schedule.Shift) error {
+	if previousShifts == nil || len(previousShifts) == 0 {
+		fmt.Printf("No previous shifts found. Starting from scratch.")
+		return nil
+	}
+
+	for i, shift := range previousShifts {
+		if _, err := time.Parse(DateFormat, shift.StartDate); err != nil {
+			return fmt.Errorf("error in previous shift entry %v, invalid value: %v, err: %v", i, shift.StartDate, err)
+		}
+		if shift.User == "" {
+			return fmt.Errorf("user cannot be empty in previous shift entry %v", i)
+		}
+	}
+
+	return nil // It's all good.
+}
+
+func (s *Scheduler) startTime(mostRecentShift *schedule.Shift) time.Time {
+	start, err := time.Parse(DateFormat, mostRecentShift.StartDate)
+	if err != nil {
+		return time.Now() // Should never happen with validation.
```

Would it make sense to return an error here then rather than return `Now`?  Or per the above comment, just return an error code from this function (which parses only the date we need) instead of pre-validating then assuming this error will always be `nil`?

If we ever do go into this block we'll probably see strange behavior as we'll have shifts that start mid-day.

---

> shifts that start mid-day

The date components are the only thing that's used, so I wouldn't predict any mid-day shifts.

Regardless, the `Schedule` function is likely to only be used once to bootstrap everything (or scrap any existing schedule and start fresh). The `ExtendSchedule` is what I intend to have invoked on a regular basis.

---

Oh I see, yeah I guess we do drop anything other than the date in the format function.

That being said, I'm generally wary of comments that say `should never happen` that do anything other than hard error...

---

Will remove when the logic gets moved to calendar sync